// graphql-to-js-client-builder/index.es.js
/*
The MIT License (MIT)
Copyright (c) 2016 Shopify Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
OR OTHER DEALINGS IN THE SOFTWARE.


Version: 1.0.0 Commit: 92d4436
*/
import { parse, visit } from 'graphql/language';
import { arrayExpression, arrowFunctionExpression, assignmentExpression, blockStatement, booleanLiteral, callExpression, expressionStatement, identifier, memberExpression, numericLiteral, objectExpression, objectProperty, program, stringLiteral, variableDeclaration, variableDeclarator } from 'babel-types';
import * as t from 'babel-types';
import generate from 'babel-generator';

// Recursive helper function for sortDefinitions
function visitFragment(fragment, fragments, fragmentsHash) {
  if (fragment.marked) {
    throw Error('Fragments cannot contain a cycle');
  }
  if (!fragment.visited) {
    fragment.marked = true;
    // Visit every spread in this fragment definition
    visit(fragment, {
      FragmentSpread(node) {
        // Visit the corresponding fragment definition
        visitFragment(fragmentsHash[node.name.value], fragments, fragmentsHash);
      }
    });
    fragment.visited = true;
    fragment.marked = false;
    fragments.push(fragment);
  }
}

// Sorts the definitions such that all fragment definitions are before operation
// definitions and fragments definitions are in reverse topological order
function sortDefinitions(definitions) {
  const fragments = definitions.filter(definition => {
    return definition.kind === 'FragmentDefinition';
  });

  // Set up a hash for quick lookup
  const fragmentsHash = {};

  fragments.forEach(fragment => {
    fragmentsHash[fragment.name.value] = fragment;
  });

  const operations = definitions.filter(definition => {
    return definition.kind === 'OperationDefinition';
  });

  const sortedFragments = [];

  fragments.forEach(fragment => {
    if (!fragment.visited) {
      visitFragment(fragment, sortedFragments, fragmentsHash);
    }
  });

  return sortedFragments.concat(operations);
}

function variableReference(operationName, variable, { variablesVar }) {
  return memberExpression(memberExpression(variablesVar, identifier(operationName)), identifier(variable.name.value));
}

function argValueToJS(argumentValue, operationName, config) {
  switch (argumentValue.kind) {
    case 'StringValue':
      return stringLiteral(argumentValue.value);
    case 'EnumValue':
      return callExpression(memberExpression(config.clientVar, identifier('enum')), [stringLiteral(argumentValue.value)]);
    case 'IntValue':
      return numericLiteral(parseInt(argumentValue.value, 10));
    case 'FloatValue':
      return numericLiteral(parseFloat(argumentValue.value));
    case 'BooleanValue':
      return booleanLiteral(argumentValue.value);
    case 'ListValue':
      return arrayExpression(argumentValue.values.map(value => {
        return argValueToJS(value, operationName, config);
      }));
    case 'ObjectValue':
      return objectExpression(argumentValue.fields.map(field => {
        return objectProperty(identifier(field.name.value), argValueToJS(field.value, operationName, config));
      }));
    case 'Variable':
      return variableReference(operationName, argumentValue, config);
    default:
      throw Error(`Unrecognized argument value type "${argumentValue.kind}"`);
  }
}

function argToJS(key, args, operationName, config) {
  const argFields = args.map(argument => {
    return objectProperty(identifier(argument.name.value), argValueToJS(argument.value, operationName, config));
  });

  return objectProperty(key, objectExpression(argFields));
}

function identifyOperation(selection, { spreadsVar }) {
  switch (selection.kind) {
    case 'Field':
      return {
        selectionConstructionArgs: [stringLiteral(selection.name.value)],
        operationMethodName: 'add',
        kind: 'Field'
      };
    case 'InlineFragment':
      return {
        selectionConstructionArgs: [stringLiteral(selection.typeCondition.name.value)],
        operationMethodName: 'addInlineFragmentOn',
        kind: 'InlineFragment'
      };
    default:
      return {
        selectionConstructionArgs: [memberExpression(spreadsVar, identifier(selection.name.value))],
        operationMethodName: 'addFragment',
        kind: 'Fragment'
      };
  }
}

function applyAlias(options, selection) {
  if (selection.alias) {
    options.push(objectProperty(identifier('alias'), stringLiteral(selection.alias.value)));
  }
}

function applyArguments(options, selection, operationName, config) {
  if (!(selection.arguments && selection.arguments.length)) {
    return;
  }

  options.push(argToJS(identifier('args'), selection.arguments, operationName, config));
}

// Returns the body of the block statement representing the selections
function selectionSetToJS(selectionSet, parentSelectionName, operationName, config) {
  const selections = selectionSet.selections.map(selection => {
    var _identifyOperation = identifyOperation(selection, config);

    const selectionConstructionArgs = _identifyOperation.selectionConstructionArgs,
          operationMethodName = _identifyOperation.operationMethodName,
          kind = _identifyOperation.kind;

    const fieldOptions = [];

    applyAlias(fieldOptions, selection);
    applyArguments(fieldOptions, selection, operationName, config);

    // Add query options (i.e. alias and arguments) to the query
    if (fieldOptions.length) {
      selectionConstructionArgs.push(objectExpression(fieldOptions));
    }

    if (['Field', 'InlineFragment'].includes(kind) && selection.selectionSet) {
      const fieldNameOrTypeConstraint = selectionConstructionArgs[0].value;

      selectionConstructionArgs.push(selectionSetToJS(selection.selectionSet, fieldNameOrTypeConstraint, operationName, config));
    }

    return expressionStatement(callExpression(memberExpression(identifier(parentSelectionName), identifier(operationMethodName)), selectionConstructionArgs));
  });

  return arrowFunctionExpression([identifier(parentSelectionName)], blockStatement(selections));
}

function fragmentVisitor(jsNodes, config) {
  return function visitor(node) {
    const selectionRootName = 'root';
    const fragmentDefinitionArguments = [stringLiteral(node.name.value), stringLiteral(node.typeCondition.name.value), selectionSetToJS(node.selectionSet, selectionRootName, null, config)];

    jsNodes.push(expressionStatement(assignmentExpression('=', memberExpression(config.spreadsVar, identifier(node.name.value)), callExpression(memberExpression(config.documentVar, identifier('defineFragment')), fragmentDefinitionArguments))));
  };
}

function referenceVariablesInOperationDeclaration(operationName, variableDefinitions, config) {
  return arrayExpression(variableDefinitions.map(variableAst => {
    return variableReference(operationName, variableAst.variable, config);
  }));
}

function operationName(definition) {
  return definition.name ? definition.name.value : '__defaultOperation__';
}

function applyName(graphQLNode, argList) {
  if (graphQLNode.name) {
    argList.push(stringLiteral(graphQLNode.name.value));
  }
}

function applyVariables(operationDefinition, argList, config) {
  const variableDefinitions = operationDefinition.variableDefinitions;

  if (!(variableDefinitions && variableDefinitions.length)) {
    return;
  }

  const name = operationName(operationDefinition);

  argList.push(referenceVariablesInOperationDeclaration(name, variableDefinitions, config));
}

function operationFactoryFunction(graphQLNode) {
  switch (graphQLNode.operation) {
    case 'query':
      return 'addQuery';
    case 'mutation':
      return 'addMutation';
    default:
      throw new Error(`Operation: "${graphQLNode.operation}" is not currently supported`);
  }
}

function operationVisitor(jsNodes, config) {
  return function visitor(node) {

    const selectionRootName = 'root';
    const operationDefinitionArgs = [];
    const name = operationName(node);

    applyName(node, operationDefinitionArgs);
    applyVariables(node, operationDefinitionArgs, config);

    operationDefinitionArgs.push(selectionSetToJS(node.selectionSet, selectionRootName, name, config));

    jsNodes.push(expressionStatement(callExpression(memberExpression(config.documentVar, identifier(operationFactoryFunction(node))), operationDefinitionArgs)));
  };
}

function typeConstraint(variableAst) {
  switch (variableAst.kind) {
    case 'NonNullType':
      return `${typeConstraint(variableAst.type)}!`;
    case 'ListType':
      return `[${typeConstraint(variableAst.type)}]`;
    default:
      return variableAst.name.value;
  }
}

function constructJSVariableDefinition(variableAst, config) {
  const variableConstructionArgs = [stringLiteral(variableAst.variable.name.value), stringLiteral(typeConstraint(variableAst.type))];

  if (variableAst.defaultValue) {
    variableConstructionArgs.push(argValueToJS(variableAst.defaultValue, null, config));
  }

  return callExpression(memberExpression(config.clientVar, identifier('variable')), variableConstructionArgs);
}

function isOperationDefinition(definition) {
  return definition.kind === 'OperationDefinition';
}

function hasVariables(definition) {
  return definition.variableDefinitions && definition.variableDefinitions.length;
}

function nameFromOperation(operationDefinition) {
  return operationDefinition.name ? operationDefinition.name.value : '__defaultOperation__';
}

function variablesDeclaration({ variablesVar }) {
  return variableDeclaration('const', [variableDeclarator(variablesVar, objectExpression([]))]);
}

function variablesHashForOperationDeclaration(operationName, { variablesVar }) {
  return expressionStatement(assignmentExpression('=', memberExpression(variablesVar, identifier(operationName)), objectExpression([])));
}

function declareVariableForOperation(operationName, variableAst, config) {
  return expressionStatement(assignmentExpression('=', memberExpression(memberExpression(config.variablesVar, identifier(operationName)), identifier(variableAst.variable.name.value)), constructJSVariableDefinition(variableAst, config)));
}

function declareVariables(graphqlAst, config) {
  const operationsWithVariables = graphqlAst.definitions.filter(isOperationDefinition).filter(hasVariables);

  if (!operationsWithVariables.length) {
    return [];
  }

  return [variablesDeclaration(config), ...operationsWithVariables.map(operationDefinition => {
    const operationName = nameFromOperation(operationDefinition);

    return [variablesHashForOperationDeclaration(operationName, config), ...operationDefinition.variableDefinitions.map(variableAst => {
      return declareVariableForOperation(operationName, variableAst, config);
    })];
  }).reduce((acc, variablesStatements) => {
    return acc.concat(variablesStatements);
  }, [])];
}

function extractFragmentDefinitons(definitions) {
  return definitions.filter(definition => definition.kind === 'FragmentDefinition');
}

function insertObjectDeclaration(nodes, identifier$$1) {
  nodes.push(variableDeclaration('const', [variableDeclarator(identifier$$1, objectExpression([]))]));
}

function declareDocument(nodes, { documentVar, clientVar }) {
  nodes.push(variableDeclaration('const', [variableDeclarator(documentVar, callExpression(memberExpression(clientVar, identifier('document')), []))]));
}

function documentToJSAst(graphQLAst, config) {
  const jsGraphQLNodes = [];

  declareDocument(jsGraphQLNodes, config);

  const sortedgraphQLAst = Object.assign({}, graphQLAst, { definitions: sortDefinitions(graphQLAst.definitions) });
  const fragmentDefinitons = extractFragmentDefinitons(sortedgraphQLAst.definitions);

  if (fragmentDefinitons.length) {
    insertObjectDeclaration(jsGraphQLNodes, config.spreadsVar);
  }

  jsGraphQLNodes.push(...declareVariables(graphQLAst, config));

  visit(sortedgraphQLAst, {
    FragmentDefinition: fragmentVisitor(jsGraphQLNodes, config),
    OperationDefinition: operationVisitor(jsGraphQLNodes, config)
  });

  return jsGraphQLNodes;
}

const defaults = {
  clientVar: 'client',
  documentVar: 'document',
  spreadsVar: 'spreads',
  variablesVar: 'variables'
};

function transformToAst(graphqlCode, {
  clientVar = 'client',
  documentVar = 'document',
  spreadsVar = 'spreads',
  variablesVar = 'variables'
} = defaults) {
  const graphQLAst = parse(graphqlCode);
  const vars = {
    clientVar: identifier(clientVar),
    documentVar: identifier(documentVar),
    spreadsVar: identifier(spreadsVar),
    variablesVar: identifier(variablesVar)
  };

  return documentToJSAst(graphQLAst, vars);
}

function transformToCode(graphqlCode, {
  clientVar = 'client',
  documentVar = 'document',
  spreadsVar = 'spreads',
  variablesVar = 'variables'
} = defaults) {
  const jsAst = transformToAst(graphqlCode, {
    clientVar,
    documentVar,
    spreadsVar,
    variablesVar
  });

  return `${generate(program(jsAst)).code}\n`;
}

export { transformToAst };export default transformToCode;
