// graphql-js-client-compiler/index.js
/*
The MIT License (MIT)
Copyright (c) 2016 Shopify Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
OR OTHER DEALINGS IN THE SOFTWARE.


Version: 0.2.0 Commit: 74777ad
*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var t = require('babel-types');
var generate = _interopDefault(require('babel-generator'));
var graphqlToJsClientBuilder = require('graphql-to-js-client-builder');
var Module = _interopDefault(require('module'));
var Client = _interopDefault(require('graphql-js-client/dev'));
var graphqlJsSchema = require('graphql-js-schema');
var graphql = require('graphql');
var graphql_utilities = require('graphql/utilities');
var graphql_language = require('graphql/language');
var path = require('path');

function baseAst(graphQlCode) {
  const clientVar = 'client';
  const documentVar = 'document';
  const spreadsVar = 'spreads';
  const variablesVar = 'variables';

  return t.functionDeclaration(t.identifier('query'), [t.identifier(clientVar)], t.blockStatement(graphqlToJsClientBuilder.transformToAst(graphQlCode, {
    clientVar,
    documentVar,
    spreadsVar,
    variablesVar
  }).concat(t.returnStatement(t.identifier(documentVar)))));
}

function compileToFunction(graphQlCode) {
  const ast = baseAst(graphQlCode);

  return `${generate(t.program([ast])).code}\n`;
}

function compileToModule(graphQlCode) {
  const ast = t.exportDefaultDeclaration(baseAst(graphQlCode));

  return `${generate(t.program([ast])).code}\n`;
}

function profileQuery(query, types) {
  return profileQueries([query], types);
}

function profileQueries(queries, types) {
  const client = new Client(types, { url: 'https://not-an-api.com' });

  const functions = queries.map(query => {
    const code = compileToFunction(query);
    const virtualModule = new Module();

    virtualModule._compile(`module.exports = ${code}`, '');

    return virtualModule.exports;
  });

  Client.resetProfiler();
  Client.startProfiling();

  functions.forEach(func => {
    func(client);
  });

  Client.pauseProfiling();

  return Client.captureProfile();
}

function schemaIDLToJson(schemaIDL) {
  const schema = graphql.buildSchema(schemaIDL);

  return graphql.graphql(schema, graphql_utilities.introspectionQuery);
}

function compileSchemaJson(schemaJson, { profile = null } = {}) {
  let schema;

  if (typeof schemaJson === 'string') {
    schema = JSON.parse(schemaJson);
  } else {
    schema = schemaJson;
  }

  return graphqlJsSchema.generateSchemaBundle(schema, 'Types', profile).then(bundle => {
    return bundle.body;
  });
}

function compileSchemaIDL(schemaIDL, { profile = null } = {}) {
  return schemaIDLToJson(schemaIDL).then(schemaJson => {
    return compileSchemaJson(schemaJson, { profile });
  });
}

function compileOptimizedSchemaJson(schemaJson, { documents }) {
  return compileSchemaJson(schemaJson).then(typesCode => {
    const typesCjsCode = typesCode.replace('export default', 'module.exports =');
    const typesModule = new Module();

    typesModule._compile(typesCjsCode, '');

    const types = typesModule.exports;
    const profile = profileQueries(documents, types);

    return compileSchemaJson(schemaJson, { profile });
  });
}

function compileOptimizedSchemaIDL(schemaIDL, { documents }) {
  return schemaIDLToJson(schemaIDL).then(schemaJson => {
    return compileOptimizedSchemaJson(schemaJson, { documents });
  });
}

function findUndefinedFragments(document) {
  const definitionNames = [];
  const spreadNames = [];

  graphql_language.visit(graphql_language.parse(document), {
    FragmentDefinition(nodes) {
      definitionNames.push(nodes.name.value);
    },
    FragmentSpread(nodes) {
      spreadNames.push(nodes.name.value);
    }
  });

  return spreadNames.filter(name => {
    return !definitionNames.includes(name);
  });
}

function fragmentFilesForDocument(documentPath, document) {
  const undefinedFragments = findUndefinedFragments(document);
  const workingDirectory = path.dirname(documentPath);

  return undefinedFragments.map(fragmentName => {
    return path.join(workingDirectory, `${fragmentName}.graphql`);
  });
}

exports.compileToFunction = compileToFunction;
exports.compileToModule = compileToModule;
exports.profileQuery = profileQuery;
exports.profileQueries = profileQueries;
exports.compileSchemaJson = compileSchemaJson;
exports.compileSchemaIDL = compileSchemaIDL;
exports.compileOptimizedSchemaJson = compileOptimizedSchemaJson;
exports.compileOptimizedSchemaIDL = compileOptimizedSchemaIDL;
exports.findUndefinedFragments = findUndefinedFragments;
exports.fragmentFilesForDocument = fragmentFilesForDocument;
